File: sdk/flutter/praxy_core/test/query_test.dart
import 'package:praxy_core/praxy_core.dart';
import 'package:test/test.dart';

const _title = Col<String>('title');
const _done = Col<bool>('done');
const _score = Col<int>('score');

void main() {
  group('Query.toJson', () {
    test('equal encodes method/attribute/values', () {
      expect(Query.equal(_title, 'Hello').toJson(), {
        'method': 'equal',
        'attribute': 'title',
        'values': ['Hello'],
      });
    });

    test('equalAny encodes multiple values under one attribute', () {
      expect(Query.equalAny(_score, [1, 2, 3]).toJson()['values'], [1, 2, 3]);
    });

    test('between encodes two values', () {
      expect(Query.between(_score, 1, 10).toJson(), {
        'method': 'between',
        'attribute': 'score',
        'values': [1, 10],
      });
    });

    test('isNull/isNotNull omit values entirely', () {
      expect(Query.isNull(_title).toJson(), {'method': 'isNull', 'attribute': 'title'});
      expect(Query.isNotNull(_title).toJson(), {'method': 'isNotNull', 'attribute': 'title'});
    });

    test('select encodes column names as values, no attribute', () {
      expect(Query.select([_title, _done]).toJson(), {
        'method': 'select',
        'values': ['title', 'done'],
      });
    });

    test('limit/offset/cursor take no attribute', () {
      expect(Query.limit(10).toJson(), {'method': 'limit', 'values': [10]});
      expect(Query.cursorAfter('abc').toJson(), {'method': 'cursorAfter', 'values': ['abc']});
    });

    test('and/or nest child query objects, not strings', () {
      final q = Query.and([Query.equal(_done, false), Query.greaterThan(_score, 5)]);
      expect(q.toJson(), {
        'method': 'and',
        'values': [
          {'method': 'equal', 'attribute': 'done', 'values': [false]},
          {'method': 'greaterThan', 'attribute': 'score', 'values': [5]},
        ],
      });
    });

    test('DateTime values encode as UTC ISO-8601', () {
      final dt = DateTime.utc(2026, 1, 2, 3, 4, 5);
      expect(Query.equal(Col<DateTime>('at'), dt).toJson()['values'], ['2026-01-02T03:04:05.000Z']);
    });

    test('encode() round-trips through jsonEncode', () {
      final encoded = Query.equal(_title, 'Hello').encode();
      expect(encoded, '{"method":"equal","attribute":"title","values":["Hello"]}');
    });

    test('raw() is a real Query value, not a string', () {
      final q = Query.raw('notBetween', attribute: 'score', values: [1, 10]);
      expect(q, isA<Query>());
      expect(q.toJson()['method'], 'notBetween');
    });
  });
}

File: sdk/js/packages/core/test/query.test.ts
import { describe, expect, it } from "vitest";
import { Col, Query } from "../src/query";

describe("Query", () => {
  it("encodes a simple filter as {method, attribute, values}", () => {
    const title = new Col<string>("title");
    expect(Query.equal(title, "Hello").toJSON()).toEqual({
      method: "equal",
      attribute: "title",
      values: ["Hello"],
    });
  });

  it("encodes limit/offset without an attribute", () => {
    expect(Query.limit(50).toJSON()).toEqual({ method: "limit", values: [50] });
    expect(Query.offset(10).toJSON()).toEqual({ method: "offset", values: [10] });
  });

  it("encodes isNull/isNotNull without values", () => {
    const deletedAt = new Col<string | null>("deletedAt");
    expect(Query.isNull(deletedAt).toJSON()).toEqual({ method: "isNull", attribute: "deletedAt" });
  });

  it("encodes and()/or() by composing child Query values, not re-parsed strings", () => {
    const status = new Col<string>("status");
    const owner = new Col<string>("owner");
    const combined = Query.or([Query.equal(status, "open"), Query.equal(owner, "me")]);
    expect(combined.toJSON()).toEqual({
      method: "or",
      values: [
        { method: "equal", attribute: "status", values: ["open"] },
        { method: "equal", attribute: "owner", values: ["me"] },
      ],
    });
  });

  it("encodes a Date value as UTC ISO-8601", () => {
    const createdAt = new Col<Date>("createdAt");
    const date = new Date("2026-01-01T00:00:00.000Z");
    expect(Query.equal(createdAt, date).toJSON()).toEqual({
      method: "equal",
      attribute: "createdAt",
      values: ["2026-01-01T00:00:00.000Z"],
    });
  });

  it("encode() returns one JSON string suitable for a repeated queries[] param", () => {
    const title = new Col<string>("title");
    expect(Query.equal(title, "Hello").encode()).toBe(
      JSON.stringify({ method: "equal", attribute: "title", values: ["Hello"] }),
    );
  });

  it("select() lists column names as values with no attribute", () => {
    const id = new Col<string>("$id");
    const title = new Col<string>("title");
    expect(Query.select([id, title]).toJSON()).toEqual({ method: "select", values: ["$id", "title"] });
  });
});
